jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCache.java

```diff
@@ -0,0 +1,241 @@
+            // TODO: Can this be avoided?
+            // Remove explicitly
+            return decorator.decorate(cache).remove(key) != null;
+         } else {
+            long duration = ttl.getDurationAmount();
+            TimeUnit timeUnit = ttl.getTimeUnit();
+            return isConditional
+                  ? decorator.decorate(cache).replace(key, oldValue, value, duration, timeUnit)
+                  : decorator.decorate(cache).replace(key, value, duration, timeUnit) != null;
+         }
+      }
+
+      return false;
+   }
+
+   protected V replace(BasicCache<K, V> cache, K key, V value) {
+      return replace(cache, key, value, IDENTITY_CACHE_DECORATOR);
+   }
+
+   protected V replace(BasicCache<K, V> cache, K key, V value, CacheDecorator decorator) {
+      checkNotNull(value, "value");
+      boolean exists = decorator.decorate(cache).containsKey(key);
+      if (exists) {
+         Duration ttl = Expiration.getExpiry(expiryPolicy, Expiration.Operation.UPDATE);
+
+         if (ttl == null || ttl.isEternal()) {
+            return decorator.decorate(cache).replace(key, value);
+         } else if (ttl.equals(Duration.ZERO)) {
+            // TODO: Can this be avoided?
+            // Remove explicitly
+            return decorator.decorate(cache).remove(key);
+         } else {
+            long duration = ttl.getDurationAmount();
+            TimeUnit timeUnit = ttl.getTimeUnit();
+            return decorator.decorate(cache).replace(key, value, duration, timeUnit);
+         }
+      }
+
+      return null;
+   }
+
+   protected boolean remove(BasicCache<K, V> cache, K key, V oldValue) {
+      V current = cache.get(key);
+      if (current != null && !current.equals(oldValue)) {
+         updateTTLForAccessed(cache, key, current);
+         return false;
+      }
+
+      return cache.remove(key, oldValue);
+   }
+
+   @Override
+   public CacheManager getCacheManager() {
+      return cacheManager;
+   }
+
+   //TODO: these were initially package-level
+   protected abstract MBeanServer getMBeanServer();
+
+   //TODO: these were initially package-level
+   protected abstract Object getCacheMXBean();
+
+   //TODO: these were initially package-level
+   protected abstract Object getCacheStatisticsMXBean();
+
+   //TODO: was package-level initially
+   public abstract void setManagementEnabled(boolean enabled);
+
+   //TODO: was package-level initially
+   public abstract void setStatisticsEnabled(boolean enabled);
+
+   protected abstract void addListener(AbstractJCacheListenerAdapter<K, V> listenerAdapter);
+
+   @Override
+   public <T> T unwrap(Class<T> clazz) {
+      return ReflectionUtil.unwrap(this, clazz);
+   }
+
+   protected static CacheDecorator IDENTITY_CACHE_DECORATOR = new CacheDecorator() {
+      @Override
+      public <K, V> BasicCache<K, V> decorate(BasicCache<K, V> cache) {
+         return cache;
+      }
+   };
+
+   protected static interface CacheDecorator {
```

Comment:
Yes, I'm using it just because withFlags() is not defined in a custom ancestor of the embedded and remote cache. The flags are cleared after the first operation is executed and for a few cases the first operation is not the one requiring the flag.

I've considered the option you're describing but creating a new cache manager and managing it's lifecycle didn't seem right. I'll have another look at this option though as force return seems to be the only flag used.